sklearn/utils/metadata_requests.py

```diff
@@ -0,0 +1,1106 @@
+"""
+Metadata Routing Utility
+"""
+
+# Author: Adrin Jalali <adrin.jalali@gmail.com>
+# License: BSD 3 clause
+
+import functools
+import inspect
+from copy import deepcopy
+from enum import Enum
+from warnings import warn
+from collections import namedtuple
+from typing import Union, Optional
+from ..externals._sentinels import sentinel  # type: ignore # mypy error!!!
+from ._bunch import Bunch
+
+# This namedtuple is used to store a (mapping, routing) pair. Mapping is a
+# MethodMapping object, and routing is the output of `get_metadata_routing`.
+# MetadataRouter stores a collection of these namedtuples.
+RouteMappingPair = namedtuple("RouteMappingPair", ["mapping", "routing"])
+
+# A namedtuple storing a single method route. A collection of these namedtuples
+# is stored in a MetadataRouter.
+Route = namedtuple("Route", ["method", "used_in"])
+
+
+class RequestType(Enum):
+    """A metadata is requested either with a string alias or this enum.
+
+    .. versionadded: 1.1
+    """
+
+    # Metadata is not requested. It will not be routed to the object having the
+    # request value as UNREQUESTED.
+    UNREQUESTED = False
+    # Metadata is requested, and will be routed to the requesting object. There
+    # will be no error if the metadata is not provided.
+    REQUESTED = True
+    # Default metadata request configuration. It should not be passed, and if
+    # present, an error is raised for the user to explicitly set the request
+    # value.
+    ERROR_IF_PASSED = None
+    # this sentinel is used in `__metadata_request__*` attributes to indicate
+    # that a metadata is not present even though it may be present in the
+    # corresponding method's signature.
+    UNUSED = sentinel("UNUSED")
```

Comment:
Question: should we use `"^UNUSED"` and `"^WARN"` and `"^UNCHANGED"` instead of these sentinels so that third party developers have a slightly easier time not having to have a hard dependency on scikit-learn? Adding a `^` at the beginning of those literals guarantees no conflict with a potential metadata name.

Reply:
ping @jnothman

Reply:
Again I am surprised by the choice of `^`. I think using strings is more familiar for scikit-learn users and developers, yes

Reply:
What's a string which people are familiar with and is not a valid identifier? Would `"$WARN$"` or `"$WARN"` work? I really have no preference here as long as we choose invalid identifiers.